docs/Tracing.md

```diff
@@ -0,0 +1,194 @@
+* duration: Used for Complete events (type 'X') to encode the time taken by the event.
+
+### Visualizing and Inspecting Traces
+
+One you have a JSON trace it is simple to visualize it using Chrome:
+
+1. Browse to `chrome://tracing`
+2. Drag and drop a Trace Event file, or hit the Load button:
+3. ![](chrome-tracing.png)
+
+### Implementation Details and Examples
+
+
+**Manual Trace Events:**
+
+A Manual Trace Event is a node like any other in the function graph, and event data storage is backed by a Placeholder:
+
+```
+size_t numEvents = 2;
+Placeholder *traceBacking = module.createPlaceholder(ElemKind::Int64Ty,
+    {numEvents, 1}, "traceBacking", false);
+    
+F->createTraceEvent("functionName", "B", traceBacking, 0);
+
+// Contents of Function
+
+F->createTraceEvent("functionName", "E", traceBacking, 0);
+
+// Compile Function F as usual
+auto compiled = executionEngine.compile(F);
+```
+
+Then, to examine the events emitted for a particular run create a TraceContext with the desired verbosity, and it should be filled out with events after execution:
+
+```
+ExecutionContext ctx;
+ctx.setTraceContext(llvm::make_unique<TraceContext>(TraceLevel::STANDARD));
+
+// Run the compiled Function as usual
+executionEngine.run(ctx);
+
+std::vector<TraceEvent>& events = ctx.getTraceContext()->getTraceEvents();
+for (TraceEvent& event : events) {
+    // inspect event
+}
+```
+
+**Auto Instrumentation**:
+
+Set the autoInstrument option in the CompilationContext before compiling.
+
+```
+  CompilationContext cctx;
+  cctx.compMode = CompilationMode::Infer;
+  cctx.backendOpts.autoInstrument = true;
+  backend->compile(F, cctx.backendOpts);
+```
+
+On a Backend that supports auto instrumentation this will decorate or configure the executable code such that Trace Events are inserted for each operator.
+
+Events are emitted in the same way as for Manual events, in the TraceContext included in a run's ExecutionContext.
+
+**Writing Trace Events to disk:**
+
+A **TraceContext** can produce JSON files suitable to be visualized in chrome via the `dumpTraceEvents` method:
+
+```
+std::vector<TraceEvent>& events = ctx.getTraceContext()->getTraceEvents();
+TraceEvent::dumpTraceEvents(events, "glow-trace.json", "glow");
+```
+
+This will output a **glow-trace.json** file in the current directory, where the process name is set to “glow”. The **tracing-compare** example in `glow/examples/` demonstrates this flow running multiple devices of different types.
+
+In Onnxifi this can be configured to always dump Trace Events via the command line flag: `--glow-dump-debug-traces`.
```

Comment:
It's not possible to set this flag via command line yet, right?
One needs to change that manually in HostManagerOnnxifi.cpp code.

Reply:
It is possible yes, that flag should work if you're using Onnxifi.

Reply:
technically it's possible but ONNXIFI does not really work well with llvm flags now.

Reply:
We have a gflag with the same name as well, try it and see :)